source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,489 @@
+pub struct SdCardClient {
+    handle: KernelHandle<SdmmcService>,
+    reply: Reusable<Envelope<Result<Response, Error>>>,
+}
+
+/// The different types of cards
+#[derive(Debug, PartialEq)]
+pub enum CardType {
+    /// Standard capacity v1
+    SD1,
+    /// Standard capacity v2 (or later)
+    SD2,
+    /// High capacity
+    SDHC,
+}
+
+/// Card status in R1 response format
+#[derive(Debug, PartialEq)]
+pub struct CardStatus(u32);
+
+/// Card identification register in R2 response format
+///
+/// | Field                 | Bits        |
+/// |:----------------------|------------:|
+/// | Manufacturer ID       | `[127:120]` |
+/// | OEM/Application ID    | `[119:104]` |
+/// | Product name          | `[103:64]`  |
+/// | Product revision      | `[63:56]`   |
+/// | Product serial number | `[55:24]`   |
+/// | Reserved              | `[23:20]`   |
+/// | Manufacturing date    | `[19:8]`    |
+/// | CRC7 checksum         | `[7:1]`     |
+/// | Not used, always 1    | `[0:0]`     |
+pub struct CardIdentification(u128);
+
+/// Published RCA in R6 response format
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct RelativeCardAddress(u32);
+
+impl SdCardClient {
+    /// Obtain an `SdCardClient`
+    ///
+    /// If the [`SdmmcService`] hasn't been registered yet, we will retry until it
+    /// has been registered.
+    pub async fn from_registry(
+        kernel: &'static Kernel,
+    ) -> Result<Self, registry::ConnectError<SdmmcService>> {
+        let handle = kernel.registry().connect::<SdmmcService>(()).await?;
+
+        Ok(Self {
+            handle,
+            reply: Reusable::new_async().await,
+        })
+    }
+
+    /// Obtain an `SdCardClient`
+    ///
+    /// Does NOT attempt to get an [`SdmmcService`] handle more than once.
+    ///
+    /// Prefer [`SdCardClient::from_registry`] unless you will not be spawning one
+    /// around the same time as obtaining a client.
+    pub async fn from_registry_no_retry(
+        kernel: &'static Kernel,
+    ) -> Result<Self, registry::ConnectError<SdmmcService>> {
+        let handle = kernel.registry().try_connect::<SdmmcService>(()).await?;
+
+        Ok(Self {
+            handle,
+            reply: Reusable::new_async().await,
+        })
+    }
+
+    async fn cmd(&mut self, cmd: Command) -> Result<Response, Error> {
+        self.handle
+            .request_oneshot(cmd, &self.reply)
+            .await
+            .map_err(|error| {
+                tracing::warn!(?error, "failed to send request to SD/MMC service");
+                Error::Other // TODO
+            })
+            .and_then(|resp| resp.body)
+    }
+
+    /// Reset the card
+    pub async fn reset(&mut self) -> Result<(), Error> {
+        self.cmd(Command::default()).await.map(|_| ())
+    }
+
+    /// Initialize the card
+    pub async fn initialize(&mut self) -> Result<CardType, Error> {
+        /// Request switch to 1.8V
+        #[allow(dead_code)]
```

Comment:
where's the dead code here?

Reply:
The `OCR_S18R` constant is currently not used, but I wanted to keep the option open to start using it in the future without having to look up the value again / forgetting it exists :sweat_smile:.